surface/include/pcl/surface/mls.h

```diff
@@ -49,32 +49,163 @@
 #include <pcl/surface/eigen.h>
 #include <pcl/surface/processing.h>
 #include <map>
+#include <boost/function.hpp>
 
 namespace pcl
 {
+
+  const double MLS_PROJECTION_CONVERGENCE_TOLERANCE = 1e-8;
+  const double MLS_MINIMUM_PRINCIPLE_CURVATURE = 1e-5;
+
   /** \brief Data structure used to store the results of the MLS fitting */
   struct MLSResult
   {
     MLSResult () : num_neighbors (0), curvature (0.0f), valid (false) {}
 
-    MLSResult (const Eigen::Vector3d &a_mean,
+    MLSResult (const Eigen::Vector3d &a_query_point,
+               const Eigen::Vector3d &a_mean,
                const Eigen::Vector3d &a_plane_normal,
                const Eigen::Vector3d &a_u,
                const Eigen::Vector3d &a_v,
-               const Eigen::VectorXd a_c_vec,
+               const Eigen::VectorXd &a_c_vec,
                const int a_num_neighbors,
-               const float &a_curvature);
+               const float &a_curvature,
+               const int a_order,
+               const bool a_polynomial_fit);
 
+    Eigen::Vector3d query_point;  /**< \brief The query point about which the mls surface was generated */
     Eigen::Vector3d mean;         /**< \brief The mean point of all the neighbors. */
     Eigen::Vector3d plane_normal; /**< \brief The normal of the local plane of the query point. */
     Eigen::Vector3d u_axis;       /**< \brief The axis corresponding to the u-coordinates of the local plane of the query point. */
     Eigen::Vector3d v_axis;       /**< \brief The axis corresponding to the v-coordinates of the local plane of the query point. */
     Eigen::VectorXd c_vec;        /**< \brief The polynomial coefficients Example: z = c_vec[0] + c_vec[1]*v + c_vec[2]*v^2 + c_vec[3]*u + c_vec[4]*u*v + c_vec[5]*u^2 */
     int num_neighbors;            /**< \brief The number of neighbors used to create the mls surface. */
     float curvature;              /**< \brief The curvature at the query point. */
+    int order;                    /**< \brief The order of the polynomial used if polynomial_fit = true */
+    bool polynomial_fit;          /**< \brief If True, a polynomial surface was fitted to the neighbors as the mls surface */
     bool valid;                   /**< \brief If True, the mls results data is valid, otherwise False. */
+
+  };
+
+  /** \brief Data structure used to store the MLS polynomial partial derivatives */
+  struct PolynomialPartialDerivative
+  {
+    double z;     /**< @brief The z component of the polynomial evaluated at z(u, v). */
+    double z_u;   /**< @brief The partial derivative dz/du. */
+    double z_v;   /**< @brief The partial derivative dz/dv. */
+    double z_uu;  /**< @brief The partial derivative d^2z/du^2. */
+    double z_vv;  /**< @brief The partial derivative d^2z/dv^2. */
+    double z_uv;  /**< @brief The partial derivative d^2z/dudv. */
```

Comment:
I agree with the `unsigned` recommendation, it expresses the intent about the expected valid range and semantics better. However, I don't think that using a smaller int will add anything to the picture and make the code more expressive. From the storage viewpoint it will probably also make no difference due to alignment.

That being said, I wonder if this variable is needed at all. Isn't it the same as `c_vec.length()`?